qualcommax: ipq807x: add support for Netgear RAX120v2 (replaces #11830)

boretom opened the pull request:
ipq807x: add support for Netgear RAX120v2

Netgear Nighthawk RAX120v2 AX WIFI router with 5 1G and 1 5G ports.

The majority of the code is based on @jewwest's PR #11830.

Specifications:
    * CPU: Qualcomm IPQ8074 Quad core Cortex-A53 2.2GHz
    * RAM: 1024MB of DDR3 (Nanya NT5CC256M16EP-EK × 2)
    * Flash: SPI-NAND 512 MiB (Winbond W29N04GZBIBA)
    * Ethernet: 4x 10/100/1000 Mbps LAN,
                1x 10/100/1000 Mbps WAN (Qualcomm QCA8075),
                1x 10/100/1000/2500/5000 Mbps LAN/WAN (Aquantia AQR111B0 PHY)
    * Wi-Fi:
        * 2.4 GHz: Qualcomm QCN5024 4x4
        * 2x 5 GHz: Qualcomm QCN5054 4x4
    * USB: 2x USB 3.0
    * LEDs: Power, 2.4GHz & 5GHz Radio, WPS, WAN, USB1 & USB2, 5G LAN
    * Keys: LEDs On/Off, Power, Reset, RFKILL, WPS
    * UART: Marked J9003 VCC TX RX GND, beginning from "1". 3.3v, 115200n8
    * Power: 19 VDC, 3.1 A

Installation:
    * Flashing OpenWrt is done in two steps:
        a) Flash *-squashfs-web-ui-factory.img from stock UI (thanks to @wangyu-).
            This writes an initramfs based OpenWrt image onto the RAX120v2
        b) From OpenWrt flash the *-squashfs-sysupgrade.bin using LuCI or the commandline

    * U-Boot allows booting an initramfs image via TFTP:
    - Set ip of your PC to 192.168.1.100
        - At the serial console interrupt boot at "Hit any key to stop autoboot:"
        - In u-boot run `tftpsrv`
        - On your PC send the OpenWrt initramfs image:

          tftp 192.168.1.1 -m binary -c put openwrt-ipq807x-generic-netgear_rax120v2-initramfs-uImage.itb

Make 5G Aquantia phy work:
    For the 5G port labeled 'lan5' to work a firmware is needed. This can be loaded in
    u-boot by writing the firmware to the correct mtd partition.

    The firmware file found in the Netgear stock firmware under /lib/firmware/ named
    'AQR-G3_v4.3.C-AQR_DNI_DR-EQ35AX8-R-prov1_ID23888_VER1311.cld' is needed and has to
    be converted to a MBN file.

    The `mkheader.py` script used here can be found in the Netgear V1.2.8.40 GPL source,
    under 'git_home/u-boot.git/tools/mkheader.py'

    Convert the CLD file to MBN using:
    $ python2 mkheader.py 0x44000000 0x13 <*.cld file> aqr_4.3.C.mbn

    This MBN file can then be flashed to the MTD partition to be used by u-boot.

    The necessary files can also be found in
        https://github.com/boretom/openwrt-fork/tree/rax120v2/aquantia-firmware

    * Write MBN file to MTD partition to be loaded automatically by u-boot:

      U-boot automatically tries to load the firmware from nand at address 0x7e00000 which
      corresponds to `/dev/mtd25` in OpenWrt.

        - find ETHPHYFW partition while running OpenWrt (expected: /dev/mtd25)

          $ fgrep -i 'ethphyfw' /proc/mtd
          mtd25: 00080000 00020000 "ethphyfw

        - copy mbn file to /tmp/ folder of the router

          $ scp aqr-v4.3.C.mbn 192.168.1.1:/tmp/

        - write mbn file to ethphyfw partition

          $ mtd write /tmp/aqr_v4.3.C.mbn /dev/mtd25

Revert to stock firmware:
    * Flash the stock firmware to the bootloader using TFTP/NMRP.

References to RAX120v2 GPL source:
    https://www.downloads.netgear.com/files/GPL/RAX120-V1.2.8.40_gpl_src.zip

Signed-off-by: Thomas Kupper <thomas.kupper@gmail.com>

boretom commented:
Because we got no answers from @jewwest since mid January I open a new PR for Netgear RAX120v2.

This PR has incorporated the following changes requested in @jewwest's PR:

- rebased to OpenWrt master of 2023-04-30

DTS:
- indentation
- newlines removed
- sort in alphabetical order
- partition labeling and addressing
- remove redundant entries
- rename dp6 to dp6_sync
- rename AQN/aqn to AQR/aqr (Aquantia 5G port related)

Not touched yet:
- robimarko: dts / g762_clk: fixedclk - Does this oscillator actually exist on the board?
- robimarko: NVMEM and ethernet entries in DTS (I removed the entries)
- documentation: Upload of Aquantia firmware using linux usermode tool
- LED/Key related stuff

Add/Fix:
- dts: add a 'ethphyfw' partition; was able to successfully write Aquantia firmware to it (see git log how-to)
- dts: change bootargs in DTS so boots for me
- dts: set all partitions not used by OpenWrt to read-only
- add u-boot envtools configuration
- image creation: remove unneeded stuff and add factory image with Netgear related stuff (netgear-dni)

musashino205 commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-rax120v2.dts:
```diff
@@ -0,0 +1,522 @@
+			linux,code = <KEY_RFKILL>;
+		};
+
+        wps {
+			label = "wps";
+			gpios = <&tlmm 57 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+
+        reset {
+            label = "reset";
+			gpios = <&tlmm 54 GPIO_ACTIVE_LOW>;
+            linux,code = <KEY_RESTART>;
+        };
+	};
```
wrong indentations

boretom replied:
Your are right @musashino205, thank you. I managed to mess it up again :|. I'll correct it.

f00b4r0 commented:
```
Signed-off-by: jewwest <8332794+jewwest@users.noreply.github.com>
[DTS & image creation correction]
Signed-off-by: Thomas Kupper <thomas.kupper@gmail.com>
```

There is no point in having a signed off with an alias and a noreply email: this will be rejected: it refers to nobody. You can probably mention the original PR in the commit message though.

Ansuel commented:
> ```
> Signed-off-by: jewwest <8332794+jewwest@users.noreply.github.com>
> [DTS & image creation correction]
> Signed-off-by: Thomas Kupper <thomas.kupper@gmail.com>
> ```
> 
> There is no point in having a signed off with an alias and a noreply email: this will be rejected: it refers to nobody. You can probably mention the original PR in the commit message though.

yep just mention in the commit and drop the sob... only way saddly...

Ansuel commented:
@robimarko can you take a look? Wonder if we can manage to merge this before the branch

robimarko commented:
Sure, It was on my TODO.

Can you please squash the commits so its easier to review?

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-rax120v2.dts:
```diff
@@ -0,0 +1,522 @@
+	label = "lan5";
+};
+
+
```
Extra newline.

Also, how are the interface MAC-s being set?

boretom replied:
The follow was in @jewwest DTS and I did remove it (and added it again now). But only dp1 to dp4 are set correct.

```
 24         ethernet0 = &dp1;
 25         ethernet1 = &dp2;
 26         ethernet2 = &dp3;
 27         ethernet3 = &dp4;
 28         ethernet4 = &dp5;
 29         ethernet5 = &dp6_syn;
```

The MACs are stored in u-boot env variables:

```
eth1addr=6c:cd:d6:c0:7b:cd
eth2addr=6c:cd:d6:c0:7b:ce
eth3addr=70:9f:1f:77:27:28
eth4addr=35:30:30:32:32:34
eth5addr=35:38:36:4b:30:31
ethact=eth0
ethaddr=6c:cd:d6:c0:7b:cc
```

The relevant boot log is:

```
[    3.527317] ssdk_switch_device_num_init[1335]:INFO:ess-switch dts node number: 1
[    3.527422] ssdk_dt_get_switch_node[1124]:INFO:ess-switch DT exist!
[    3.533813] ssdk_dt_parse_access_mode[999]:INFO:switch_access_mode: local bus
[    3.539780] ssdk_dt_parse_access_mode[1012]:INFO:switchreg_base_addr: 0x3a000000
[    3.547075] ssdk_dt_parse_access_mode[1013]:INFO:switchreg_size: 0x1000000
[    3.554540] ssdk_dt_parse_mac_mode[327]:INFO:mac mode = 0x0
[    3.561216] ssdk_dt_parse_mac_mode[336]:INFO:mac mode1 = 0xff
[    3.566689] ssdk_dt_parse_mac_mode[345]:INFO:mac mode2 = 0xd
[    3.572634] ssdk_dt_parse_uniphy[364]:INFO:ess-uniphy DT exist!
[    3.578412] ssdk_dt_parse_intf_mac[942]:INFO:dp1 MAC 6c:cd:d6:c0:7b:cc
[    3.583995] ssdk_dt_parse_intf_mac[942]:INFO:dp2 MAC 6c:cd:d6:c0:7b:cd
[    3.590587] ssdk_dt_parse_intf_mac[942]:INFO:dp3 MAC 6c:cd:d6:c0:7b:ce
[    3.597102] ssdk_dt_parse_intf_mac[942]:INFO:dp4 MAC 70:9f:1f:77:27:28
...
[    5.071711] Generic PHY 90000.mdio-1:00: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:00, irq=POLL)
[    5.075836] nss-dp 3a001000.dp1 lan4: Registered netdev lan4(qcom-id:1)
[    5.085208] Generic PHY 90000.mdio-1:01: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:01, irq=POLL)
[    5.091960] nss-dp 3a001200.dp2 lan3: Registered netdev lan3(qcom-id:2)
[    5.101364] Generic PHY 90000.mdio-1:02: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:02, irq=POLL)
[    5.108074] nss-dp 3a001400.dp3 lan2: Registered netdev lan2(qcom-id:3)
[    5.117498] Generic PHY 90000.mdio-1:03: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:03, irq=POLL)
[    5.124191] nss-dp 3a001600.dp4 lan1: Registered netdev lan1(qcom-id:4)
[    5.133519] GMAC5(ffffff8002d7d8c0) Invalid MAC@ - using e2:13:3c:1a:59:eb
[    5.139919] Generic PHY 90000.mdio-1:04: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:04, irq=POLL)
[    5.147267] nss-dp 3a001800.dp5 wan: Registered netdev wan(qcom-id:5)
[    5.156540] GMAC6(ffffff8002e208c0) Invalid MAC@ - using ca:18:79:26:23:33
[    5.163085] Generic Clause 45 PHY 90000.mdio-1:07: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:07, irq=POL)
[    5.170083] nss-dp 3a007000.dp6-syn lan5: Registered netdev lan5(qcom-id:6)
```

robimarko replied:
If they are stored in U-boot env then those can be used via NVMEM to populate MAC-s, cause having random MAC-s doesn't really sound normal.

boretom replied:
Maybe it was that part of the [discussion from @jewwest PR](https://github.com/openwrt/openwrt/pull/11830#discussion_r1081879908)?

> @jewwest : I think we started chatting about this before. SMEM provides the partition layout represented by QCA however this is incomplete. Netgear overrides it by hard coding their partition layout in ofpart.c that was used as a foundation for these entries.

Can I still use NVRAM?

But I'll try again, I did try using NVMEM similar to how 'ipq8074-wxr-5950ax12' or 'ipq8074-nbg7815.dts' or 'ipq8072-wax218.dts use it but was not successful.

boretom replied:
Gave NVMEM a go and it didn't work. It seems someone is not happy about the MACs that start with '35:30:30' and '35:38:36'. According to MAC lookup they are not assigned (but that seems to be true for the '70:9f:1f' one, too).

All the MACs are saved in the 'boarddata1' partition, starting from offset 0. I replaced the two '35:...' MACs with valid one and then the interfaces get theses MACs successfully. And it works then with both methods, using ethernetX aliases and NVRAM.

So I don't know how to resolve that. How to generate unique MACs and have them assigned in another way?

robimarko replied:
Those `35:..` MAC-s are probably invalid so they are rejected.

So, stock FW assigns random MAC-s on each boot?

boretom replied:
Yes, in stock firmware it does for me if I run `ethtool -P eth4` or `ethtool -P eth5`, after every reboot I get a new pair of MACs.

`arp` on my connecting client or uplink report the same MAC for all the LAN and for WAN it's "LAN's MAC + 1"

robimarko replied:
But the rest are static?

If so, then at least those should be set, ideally using NVMEM.

boretom replied:
Yes the rest is static and they all are taken from the 'boardata1' partition.

NVMEM: Okey, but that adds a few lines of code and doesn't seem to behave different then when I set alias. The alias methods picked up the MAC addresses change in the boarddata1 partition. Is the aliases method not recommended?

E.g. I add:

```
 &dp1 {
        status = "okay";
        phy-handle = <&qca8075_0>;
        label = "lan4";
+       nvmem-cells = <&macaddr_dp1>;
+       nvmem-cell-names = "mac-address";
 };

 &dp2 {
        status = "okay";
        phy-handle = <&qca8075_1>;
        label = "lan3";
+       nvmem-cells = <&macaddr_dp2>;
+       nvmem-cell-names = "mac-address";
 };
[and so on ...]
                                label = "boarddata1";
                                reg = <0x1180000 0x0100000>;
                                read-only;
+
+                               compatible = "nvmem-cells";
+                               #address-cells = <1>;
+                               #size-cells = <1>;
+
+                               macaddr_dp1: macaddr@0 {
+                                       reg = <0x0 0x6>;
+                               };
+
+                               macaddr_dp2: macaddr@1 {
+                                       reg = <0x6 0x6>;
+                               };
[and so on ...]
```

That works but as said the result is the same as with aliases.

boretom replied:
And since the 'wan' interface is one of the two interfaces with invalid MAC: How can it made sure that the WAN interface gets a correct permanent MAC? Because I assume an ever changing MAC won't be going down well with PPPoE and other services.

boretom replied:
@robimarko do you have an idea about the changing MAC addresses, especially for the WAN interface?

robimarko replied:
Well, if the stock FW does the same and there is no MAC allocated for those not much can be done, we cant invent a MAC for them, its a shitty move from the vendor honestly.

boretom replied:
I see; in stock although the two MACs are generated after every reboot; since they create a bridge for WAN too the MAC showing up in ARP is "LAN-base-MAC + 1". Not sure if that behavior could (or should) be replicated in OpenWrt.

boretom commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-rax120v2.dts:
```diff
@@ -0,0 +1,522 @@
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <32768>;
+	};
```
It's the clock for the fan and in the source code from Netgear, in file `git_home/linux.git/arch/arm64/boot/dts/qcom/qcom-ipq807x-soc.dtsi` I find:

```
1066             fan_left: g761@3e {
1067                 compatible = "gmt,g761";
1068                 reg = <0x3e>;
1069                 clocks =<&g761_clk>;
1070                 fan_gear_mode = <0>;
1071                 fan_starty = <1>;
1072                 pwm_polarity = <0>;
1073             };
...
2118         g761_clk: g761-oscillator{
2119               compatible = "fixed-clock";
2120               #clock-cells = <0>;
2121               clock-frequency = <32768>;
2122
2123         };
```

@jewwest used the same values but 'compatible = "gmt,g762"; instead of "compatible = "gmt,g761"; like in the Netgear sources.

I never heard the fan running so not sure if 'g762' is correct or if the fan code is already part of the soc code.

boretom replied:
I will use the SOC's sleep clock and see how I can verify that it work.

I really thought that the vendor DTS is a hint. But since I (almost certainly) converted my first RAX120v2 into a paper weight I can look at the IC on the PCP.

boretom replied:
Good Morning @robimarko, I had a look at the board and right next to the fan connector is a 'W51S G761' IC.

![Netgear-RAX120v2-fan](https://user-images.githubusercontent.com/52218/236743221-2934290c-7bcb-4df3-b6fe-a7c096e9f4e9.jpeg)

robimarko replied:
You can't really trust the vendor's DTS for anything, they will just go the path of least resistance no matter how broken it is.

If it's G761 then `gmt,g763` should be used as that seems to be the single-channel as well, though kernel driver makes no difference between them.

boretom replied:
Ok, but do I keep the fan section and refer to the sleep clock? And how do I refer to the sleep clock?

right now:

```
296 &blsp1_i2c2 {
297     status = "okay";
298
299     fan: g762@3e {
300         compatible = "gmt,g762";
301         reg = <0x3e>;
302         clocks =<&g762_clk>;
303         fan_gear_mode = <0>;
304         fan_start = <1>;
305         pwm_polarity = <0>;
306     };
307 };
```

robimarko replied:
Yes obviously, but drop the fan label if that is not being referenced on and for the clock its just:
`clocks =<&sleep_clk>;`

boretom commented:
- Commit squash so there is only one commit
- remove @jewwest sob
- add remark in commit message that this code is based on @jewwest's code

mrkiko commented:
Hi!!
Thanks a lot for your work!

A note on the commit instruction: at some point you mention the .cld file as needed, but then use the .mbn one only. Am I missing something there?

And a curiosity of my own: how does NMRP protocol work? If it actually does, what does stop us from flashing OpenWrt via NMRP? I ask because, of course, being able to flash OpenWrt without opening the device would be nice. :)

Thanks!

boretom commented:
@mrkiko : 

> A note on the commit instruction: at some point you mention the .cld file as needed, but then use the .mbn one only. Am I missing something there?

You're right, that is still a todo. The CLD file is the firmware file that can be used to upload using a linux usermode tool. The MBN is a converted CLD file so u-boot can load it directly from a MTD device. I'll clarify that, the instructions for CLD already exists in the OpenWrt entry for RAX120v2. But I was unsure how long a commit message should get.

> And a curiosity of my own: how does NMRP protocol work? If it actually does, what does stop us from flashing OpenWrt via NMRP? I ask because, of course, being able to flash OpenWrt without opening the device would be nice. :)

The thing stopping us from flashing without UART access is that you have to modify the 'bootcmd' u-boot variable and for that it may be enough to get telnet access to the router (or maybe it's possible to include it in the flashing process). 

BUT someone with a pristine router would have to check if the 'appsblenv' partition is empty. U-boot does show the correct env variables even if the 'appsblenv' partition is empty. I don't know where it gets them from, using `hexdump` I did try to find like the mac address in the ART partition and other was unable to find it yet.
If the partition is empty that would be a problem because then `fw_setenv` uses the default env which is not correct at all.

A long as you leave the `nmrp` call in the 'bootcmd' you should be able to flash back to stock or OpenWrt. But it's tested only by me so far and I may have to add something since from time to time it skips over `nmrp` right now.

@musashino205 : Thank you (and for letting me know how to change it), I was wondering why @jewwest icon showed up.

boretom committed:
ipq807x: add support for Netgear RAX120v2

Netgear Nighthawk RAX120v2 AX WIFI router with 5 1G and 1 5G ports.

The majority of the code is based on @jewwest's PR #11830.

Specifications:
    * CPU: Qualcomm IPQ8074 Quad core Cortex-A53 2.2GHz
    * RAM: 1024MB of DDR3 (Nanya NT5CC256M16EP-EK × 2)
    * Flash: SPI-NAND 512 MiB (Winbond W29N04GZBIBA)
    * Ethernet: 4x 10/100/1000 Mbps LAN,
                1x 10/100/1000 Mbps WAN (Qualcomm QCA8075),
                1x 10/100/1000/2500/5000 Mbps LAN/WAN (Aquantia AQR111B0 PHY)
    * Wi-Fi:
        * 2.4 GHz: Qualcomm QCN5024 4x4
        * 2x 5 GHz: Qualcomm QCN5054 4x4
    * USB: 2x USB 3.0
    * LEDs: Power, 2.4GHz & 5GHz Radio, WPS, WAN, USB1 & USB2, 5G LAN
    * Keys: LEDs On/Off, Power, Reset, RFKILL, WPS
    * UART: Marked J9003 VCC TX RX GND, beginning from "1". 3.3v, 115200n8
    * Power: 19 VDC, 3.1 A

Installation:
    * Flashing OpenWrt is done in two steps:
        a) Flash *-squashfs-web-ui-factory.img from stock UI (thanks to @wangyu-).
           This writes an initramfs based OpenWrt image onto the RAX120v2
        b) From OpenWrt flash the *-squashfs-sysupgrade.bin using LuCI or the commandline

    * U-Boot allows booting an initramfs image via TFTP:
    - Set ip of your PC to 192.168.1.100
        - At the serial console interrupt boot at "Hit any key to stop autoboot:"
        - In u-boot run `tftpsrv`
        - On your PC send the OpenWrt initramfs image:

          tftp 192.168.1.1 -m binary -c put openwrt-ipq807x-generic-netgear_rax120v2-initramfs-uImage.itb

Make 5G Aquantia phy work:
    For the 5G port labeled 'lan5' to work a firmware is needed. This can be loaded in
    u-boot by writing the firmware to the correct mtd partition.

    The firmware file found in the Netgear stock firmware under /lib/firmware/ named
    'AQR-G3_v4.3.C-AQR_DNI_DR-EQ35AX8-R-prov1_ID23888_VER1311.cld' is needed and has to
    be converted to a MBN file.

    The `mkheader.py` script used here can be found in the Netgear V1.2.8.40 GPL source,
    under 'git_home/u-boot.git/tools/mkheader.py'

    Convert the CLD file to MBN using:
    $ python2 mkheader.py 0x44000000 0x13 <*.cld file> aqr_4.3.C.mbn

    This MBN file can then be flashed to the MTD partition to be used by u-boot.

    The necessary files can also be found in
        https://github.com/boretom/openwrt-fork/tree/rax120v2/aquantia-firmware

    * Write MBN file to MTD partition to be loaded automatically by u-boot:

      U-boot automatically tries to load the firmware from nand at address 0x7e00000 which
      corresponds to `/dev/mtd25` in OpenWrt.

        - find ETHPHYFW partition while running OpenWrt (expected: /dev/mtd25)

          $ fgrep -i 'ethphyfw' /proc/mtd
          mtd25: 00080000 00020000 "ethphyfw

        - copy mbn file to /tmp/ folder of the router

          $ scp aqr-v4.3.C.mbn 192.168.1.1:/tmp/

        - write mbn file to ethphyfw partition

          $ mtd write /tmp/aqr_v4.3.C.mbn /dev/mtd25

Revert to stock firmware:
    * Flash the stock firmware to the bootloader using TFTP/NMRP.

References to RAX120v2 GPL source:
    https://www.downloads.netgear.com/files/GPL/RAX120-V1.2.8.40_gpl_src.zip

Reviewed-by: Robert Marko <robimarko@gmail.com>
Signed-off-by: Thomas Kupper <thomas.kupper@gmail.com>

mrkiko commented:
Thanks for your great work. Where you ableto recover your first unit?
Anything that can be done to help out?

robimarko commented:
I am seeing now that AQR PHY-s are identified as generic, what is the exact model that is being used?

boretom commented:
> I am seeing now that AQR PHY-s are identified as generic, that is an issue as AQR-s use a lot of vendor-specific registers. What is the exact model that is being used?

If I read the pictures on [RAX120v2 wiki dev ru](https://wikidevi.wi-cat.ru/Netgear_RAX120_(Nighthawk_AX12)) correctly it's the `AQR111-B0-C`. It doesn't show up when I run `mdio 900*`

robimarko commented:
Its a C45 PHY, so it has not C22 ID so the in mdio tools will be 0.
You can read the ID using mdio tools like this:
`mdio 9* mmd 0:1 2`
`mdio 9* mmd 0:1 3`

Just replace the 0 in `0:1` with the MDIO AQR PHY address.

boretom commented:
> Its a C45 PHY, so it has not C22 ID so the in mdio tools will be 0. You can read the ID using mdio tools like this: `mdio 9* mmd 0:1 2` `mdio 9* mmd 0:1 3`

Now I remember, you explained it to me before. It is the AQR111-B0 in that case:

```
$ mdio 9000* mmd 7:1 2; mdio 9000* mmd 7:1 3
0x03a1
0xb612
```

boretom commented:
2023-05-21: github PR change log:

- rebase to master (2023-05-21, commit e2f9fa42044a2660f702a9b51b14cbde24a13702)
- change git author to me (thanks @musashino205)
- update git commit message to include Keys, LEDs and clarification about the Aquantia firmware files
- change fan clock to use existing sleep clock, remove unused clock & fan label
- dts: use nvmem to set all six MACs, lan5 and wan still change after every reboot
- change git commit message regarding flashing: use the initramfs image to flash the sysupgrade image.
  No changes to u-boot env variables necessary anymore.
- flashing using the stock GUI is not working right now. The firmware is accepted and uploaded but
  rootfs can't be found afterwards.

robimarko commented:
> > Its a C45 PHY, so it has not C22 ID so the in mdio tools will be 0. You can read the ID using mdio tools like this: `mdio 9* mmd 0:1 2` `mdio 9* mmd 0:1 3`
> 
> Now I remember, you explained it to me before. It is the AQR111-B0 in that case:
> 
> ```
> $ mdio 9000* mmd 7:1 2; mdio 9000* mmd 7:1 3
> 0x03a1
> 0xb612
> ```

Perfect, support for it should be added to the Aquantia driver so it actually properly works, the generic C45 driver is not enough.

robimarko commented on target/linux/ipq807x/image/generic.mk:
```diff
@@ -71,6 +71,30 @@ define Device/edimax_cax1800
+	IMAGE/sysupgrade.bin := append-kernel | pad-offset $$$$(BLOCKSIZE) 64 | \
+		append-uImage-fakehdr filesystem | sysupgrade-tar kernel=$$$$@ | \
+		append-metadata
+	IMAGE/factory.img := append-kernel | pad-offset $$$$(BLOCKSIZE) 64 | \
```
Does this image actually work?

boretom replied:
The factory.img not, no. It can be flashed in stock ui and the kernel runs but can't find the rootfs. sysupgrade does work for me when flashed from initramfs.

Shall I remove the factory.img part as long as it doesn't work?

boretom commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -1,6 +1,5 @@
 include $(TOPDIR)/rules.mk
 include $(INCLUDE_DIR)/version.mk
-
```
Yes, indeed ... had to manually solve a merge conflict after rebasing because a few BFD have been upstreamed (commit 8217f02a1c5e) and it seems I deleted a line to many. I'll correct it.

boretom commented:
> > Now I remember, you explained it to me before. It is the AQR111-B0 in that case:
> > ```
> > $ mdio 9000* mmd 7:1 2; mdio 9000* mmd 7:1 3
> > 0x03a1
> > 0xb612
> > ```
> 
> Perfect, support for it should be added to the Aquantia driver so it actually properly works, the generic C45 driver is not enough.

I reach a limit in my English and don't completely understand: is there already support, and are we talking upstream Aquantia driver support? Or do I have to specify the device id instead of the generic C45 in the DTS?

boretom commented:
> Not upstream as parts are missing, check how AQR113C and others have been added to OpenWrt, its just a matter of adding an ID

I see, yes. I created a patch back in January (for my own use) for the AQR111 and AQR111B0 but since I'm only a simple, untrained monkey without much knowledge I don't know if there's a need to supply custom config functions or not.

And I can't find anything about AQR111 on the netdev mailing list. Would this mailing list be the appropriate place to ask for help? Or is there a OpenWrt guy to go to?

robimarko commented:
AQR111 should basically be like AQR107 AFAIK as far as driver is concerned

boretom commented:
> AQR111 should basically be like AQR107 AFAIK as far as driver is concerned

It seems there is more work needed. Adding support the same way as for AQR112 or AQR107 makes it behave worse then without. Auto-neg on a 10G switch port doesn't work and it falls back to 1G (see message below). But at least if I set the `ethtool -s lan5 advertise 0x1000000000000` 5G works. And with or without 10Gbit is wrongly advertised.

```
# dmesg | fgrep AQR
[    5.429902] Aquantia AQR111 90000.mdio-1:07: FW 4.3, Build 12, Provisioning 1
[    5.438556] Aquantia AQR111 90000.mdio-1:07: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:07, irq=POLL)
[   19.689090] Aquantia AQR111 90000.mdio-1:07: Downshift occurred from negotiated speed 10Gbps to actual speed 1Gbps, check cabling!
```

On the plus side the driver let me know what it thinks. The driver detects it as an AQR111, not AQR111B0 although `mdio` and u-boot say it's an AQR111B0

```
NAND read: device 0 offset 0x7e00000, size 0x80000
 524288 bytes read: OK
CRC check good on phy fw file (0x359C)
PHYFW:Loading IRAM...........done.
PHYFW:Loading DRAM..............done.
phy fw image load  CRC-16 (0x6F69) does not match calculated CRC-16 (0xFE36)
PHY ID1: 0x3a1
PHY ID2: 0xb612
```

But that's just your information, it's maybe a bit off topic and has to be solved somewhere else and by someone else.

boretom commented:
rebase to latest 4c5a9da8699a to include the testing kernel config for ipq807x from @robimarko. Everything works still the same with the testing kernel.

questions:
- renaming port 'lan5' to '5g-0' to indicate that this port can be used as a lan or wan port
- including the AQR111 patch (create and tested only by me) in this PR? Or better: who do ping to have a look at it and the check the reason mdio reports 10G as supported speed.

boretom commented:
* rebase to latest https://github.com/openwrt/openwrt/commit/c4be106f4d989beae7e6dc0066c0595e1b1caa00 that includes renaming 'ipq807x' target to 'qualcommax', subtarget 'ipq807x' (@robimarko).
To be aware: 
    * if you use `scripts/diffconfig.sh` before you have to run it again now.
    * `aq-fw-download` doesn't get included right now since it depends on target `ipq807x` and not `qualcommax`
* include the AQR111 patch

The 'lan5' port is not renamed to '5g-0' yet (no feedback)

boretom commented:
rebase to latest https://github.com/openwrt/openwrt/commit/2eac3239a862f94fb5661dd4f0b0398965128695. Everything still works as before.

@robimarko are technical reason this device not yet merged? Anything to improve? And on another note, would you be able to adapt the nss-package repo to include the move of the ipq807x target to qualcommax/ipq807x?

robimarko commented:
I need to look over the PR again, I can only review it as I dont have commit rights.

`nss-packages` are a legacy thing, not maintained at all.
You can drop the instructions on using `aq-fw-download` as that nor SSDK shell is present in OpenWrt.

boretom commented:
> I need to look over the PR again, I can only review it as I dont have commit rights.

I invited you as an collaborator if you wish to do so.

> `nss-packages` are a legacy thing, not maintained at all. You can drop the instructions on using `aq-fw-download` as that nor SSDK shell is present in OpenWrt.

I see; what would the process of uploading the firmware in user space be without `aq-fw-download`?

robimarko commented:
I meant, I cannot merge your PR as I dont have commit rights for OpenWrt.

There isn't really a way outside of `aq-fw-download`,  I am currently working on FW loading in the AQR driver itself.
For now the best way is to use U-boot to load the FW before booting.

boretom commented:
> I meant, I cannot merge your PR as I dont have commit rights for OpenWrt.

I see, who would you recommend I ping?
 
> There isn't really a way outside of `aq-fw-download`, I am currently working on FW loading in the AQR driver itself. For now the best way is to use U-boot to load the FW before booting.

I'm just patching the `aq-fw-download` Makefile right now so it still works. Looking forward to FW loading from the AQR driver :), thanks.

Djfe commented:
friendly ping
@hauke

wangyu- commented:
this PR now supports to be flashed without serial access.  (an intermediate image need to be used though)

see:

 https://forum.openwrt.org/t/adding-openwrt-support-for-netgear-rax120-nighthawk-ax12/75592/274?u=wangyu

boretom commented:
rebase to latest https://github.com/openwrt/openwrt/commit/881235c713fae8692190178561af4eb2dee4ead1.
Using a 2-step flashing process UART access is _not_ needed anymore, thanks to @wangyu-!

Copy-Paste from commit message regarding the flashing process:
```
* Flashing OpenWrt is done in two steps:
    a) Flash *-squashfs-web-ui-factory.img from stock UI (thanks to @wangyu-).
        This writes an initramfs based OpenWrt image onto the RAX120v2
    b) From OpenWrt flash the *-squashfs-sysupgrade.bin using LuCI or the commandline
```

For the 5G lan5 interface to work, the only supported method right now is writing it to the ethphyfw partition (see commit log on how to). The `aq-fw-download` utility is not supported by @robimarko anymore (it can be patched to work if necessary). He's working on FW loading in the driver itself.

zsh commented:
Just can't believe after months of development, this PR still not merged yet. I believe RAX120v2 support is very worth looking forward to since the official firmware is buggy and has constant reboot issue not fixed for more than a year.

Would love to try this on my own RAX120v2.

Thanks!

robimarko commented:
LGTM, so:
`Reviewed-by: Robert Marko <robimarko@gmail.com>`

musashino205 commented:
Shouldn't the commit title be replaced with ```qualcommax: ``` (or ```qualcommax: ipq807x: ```)?

robimarko commented:
Yeah, it probably should be: `qualcommax: ipq807x:`

boretom commented:
Well, I'll change it ... I didn't think that this is a pressing issue since the tagging is correct.

hauke commented:
Please send your changes to the Aquantia driver to the netdev mailing list to get it integrated into upstream Linux kernel. I think it looks ok, it is annoying that Marvell does not take care of upstream support for their PHYs.

xabolcs commented:
Ahh, the conmit title left unchanged. :confused:

boretom commented:
> Please send your changes to the Aquantia driver to the netdev mailing list to get it integrated into upstream Linux kernel. I think it looks ok, it is annoying that Marvell does not take care of upstream support for their PHYs.

Ok, I'll send it upstream.

boretom commented:
> Ahh, the commit title left unchanged. 😕

Nope, I changed it, it reads "qualcommax: ipq807x: ...". Or what did you have in mind?

robimarko commented:
You changed the PR title but not the commit one, but we will survive.

boretom commented:
> You changed the PR title but not the commit one, but we will survive.

Oh, I see ... totally forgot that, Sorry. Search for RAX120 and IPQ807x will still bring the commit up.

boretom commented:
@hauke since sending patches upstream is a task I've only done once: I have created a patch against netdev net-next and it compiles fine. But I can't really test it with the 6.6 kernel, will that be accepted or how to I proceed?

I ran the checkpatch and the get-maintainer script and think I'd be ready from that point of view.

robimarko commented:
I use Buildroot to roll linux-next kernels for development and semi-regularly push the required stuff to get it working in:
https://github.com/robimarko/buildroot-external-ipq807x

boretom commented:
Thank you @robimarko, I'll read into how to use a different kernel (and your buildroot) give it a go next week.

boretom commented:
> I use Buildroot to roll linux-next kernels for development and semi-regularly push the required stuff to get it working in: https://github.com/robimarko/buildroot-external-ipq807x

@robimarko I now got the time to have a look at it. But to be honest I have no clue what to do with your buildroot repo. Would you be able to explain in a few step?

I don't recognize the folder structure from either OpenWrt or the kernel (not that says much).

robimarko commented:
Its an external feed for Buildroot project, I cant really write any instructions but it is quite straightforward.
You just point to it from buildroot via `BR2_EXTERNAL=path_to_external_feed` and then you can use it as part of Buildroot which is well documented.

boretom commented:
@robimarko thanks for being patient. I don't know what form of dyslexic I am but I re-read you first comment regarding it again ... I never read the Buildroot part.

I will looking into Buildroot, again thanks a lot!
